### PR TITLE
Support proto encoding

### PR DESCRIPTION
**Pull request by ganglyu**
#### Why I did it
Need to support GNMI request with proto encoding.

#### How I did it
Update gnmi client and gnmi server to support both json ietf and proto encoding.

#### How to verify it
Run GNMI unit test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: gnmi_server/server.go [hunk 1]
```diff
@@ -393,6 +394,7 @@ func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetRe
 	/* Fetch the prefix. */
 	prefix := req.GetPrefix()
 	extensions := req.GetExtension()
+	encoding := gnmipb.Encoding_JSON_IETF
```
Comment (qiluo-msft) [inline gnmi_server/server.go:397]: >encoding [](http://example.com/codeflow?start=1&length=8)

Get the encoding from Set request packet?

**Review by zbud-msft — APPROVED** [review]

**Comment by zbud-msft** [comment]
LGTM

**Comment by liuh-80** [comment]
LGTM

File: sonic_data_client/mixed_db_client.go [hunk 2]
```diff
@@ -509,6 +532,31 @@ func (c *MixedDbClient) tableData2Msi(tblPath *tablePath, useKey bool, op *strin
...
+		if ok {
+			return &gnmipb.TypedValue{
+				Value: &gnmipb.TypedValue_ProtoBytes{
+					ProtoBytes: value.([]byte),
+				}}, nil
+		} else {
+			return nil, fmt.Errorf("No proto bytes found in msi %v", msi)
+		}
+	}
+	return nil, fmt.Errorf("Unknown encoding %v", c.encoding)
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:557]: >return [](http://example.com/codeflow?start=1&length=6)

Add `else`?

File: sonic_data_client/mixed_db_client.go [hunk 3]
```diff
@@ -492,7 +506,16 @@ func (c *MixedDbClient) tableData2Msi(tblPath *tablePath, useKey bool, op *strin
 		if tblPath.jsonTableKey != "" { // If jsonTableKey was prepared, use it
 			err = c.makeJSON_redis(msi, &tblPath.jsonTableKey, op, fv)
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:507]: >makeJSON_redis [](http://example.com/codeflow?start=11&length=14)

Do you need to check protobuf encoding before makeJson?

**Comment by qiluo-msft** [comment]
> 			log.V(2).Infof("makeJSON err %s for fv %v", err, fv)

This message is directly related to makeJson, not related to protobuf.
You may move it near makeJson.

---
Refers to: sonic_data_client/mixed_db_client.go:527 in 4f582a8. [](commit_id = 4f582a829ca9d2232cc59f677f6c44fef8274fcf, deletion_comment = False)

File: sonic_data_client/mixed_db_client.go (hunk 3 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Updated

File: sonic_data_client/mixed_db_client.go (hunk 2 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Updated

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**